appcelerator/titanium_mobile
# [TIMOB-26260] Android: Updated "Ti.Android.Service" to handle Android P "FOREGROUND_SERVICE" permission

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-26260

**Summary:**
Note that if you target API Level 28 (Android 9.0) or higher, then Google now requires that you add "android.permission.FOREGROUND_SERVICE" to the "AndroidManifest.xml" file. If you don't have this permission, then a permission exception will be thrown when attempting to set up a foreground service.

This PR does not prevent an exception. Instead, it'll throw the below exception message instead to help clarify what is needed. Also, this PR will throw this exception on older Android OS versions too if the permission is missing, which Google does not do, but it is intended to help developers who don't have an Android P device to help with.
```
[Developer Error] Ti.Android.Service.foregroundNotify() requires manifest permission: android.permission.FOREGROUND_SERVICE
```

Also note that the above only applies when targeting Android 9.0 (aka: API Level 28) or higher.

**Test:**
1. Set up a test project as shown in PR #10076.
2. Set up the "tiapp.xml" to target Android API Level 27.
```xml
<uses-sdk android:minSdkVersion="16" android:targetSdkVersion="27"/>
```
3. Build and run the app on Android 9.
4. Verify that a foreground service notification appears on the status bar.
5. Set up the "tiapp.xml" to target Android API Level 28.
```xml
<uses-sdk android:minSdkVersion="16" android:targetSdkVersion="28"/>
```
6. Build and the app on Android 9.
7. Verify that a "[Developer Error]" exception appears.
8. Run the app on an Android OS older that 9.0.
9. Verify that a "[Developer Error]" exception appears.
10. Add the following Android manifest permission to the "tiapp.xml" file.
```xml
<uses-permission android:name="android.permission.FOREGROUND_SERVICE"/>
```
11. Build and run the app on an Android 9 device.
12. Verify that a foreground service notification appears on the status bar.
13. Run the app on an Android OS older that 9.0.
14. Verify that a foreground service notification appears on the status bar.
